src/Autofac/Core/Lifetime/LifetimeScope.cs

```diff
@@ -49,6 +49,7 @@ public class LifetimeScope : Disposable, ISharingLifetimeScope, IServiceProvider
         /// </summary>
         private readonly object _synchRoot = new object();
         private readonly ConcurrentDictionary<Guid, object> _sharedInstances = new ConcurrentDictionary<Guid, object>();
+        private readonly ConcurrentDictionary<(Guid, Guid), object> _sharedQualifiedInstances = new ConcurrentDictionary<(Guid, Guid), object>();
```

Comment:
Splitting the instance caches up instead of forcing both simple and qualified instances into a shared cache seems to have split the difference on performance cost. Non-decorator instances see no change, but decorators have a bit more cost in generating/comparing keys in moving to the ValueTuple